content/courses/advanced-ada/parts/data_types/records.rst

```diff
@@ -2918,20 +2928,684 @@ In this example, :ada:`Arr_Sub` and :ada:`Rec_Sub` are unconstrained subtypes.
+             when Off =>
+                null;
+             when On =>
+                Info : Device_Info;
+          end case;
+       end record;
+
+    end Devices;
+
+The :ada:`Device` type from this example has a variant part. Depending on the
+value of the :ada:`State` discriminant, it can be either a null record (when
+:ada:`State` is :ada:`Off`) or have the :ada:`Info` component (when
+:ada:`State` is :ada:`On`).
+
+Let's look at a test application for the :ada:`Devices` package:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Simple_Device
+    :class: ada-run-expect-failure
+
+    with Devices; use Devices;
+
+    procedure Show_Device is
+       D     : Device;
+       D_Off : Device (Off);
+       D_On  : Device (On);
+    begin
+       D := D_Off;
+       --  OK!
+
+       D := D_On;
+       --  OK!
+
+       D_Off := D_On;
+       --       ^^^^
+       --  CONSTRAINT_ERROR!
+
+       D_On  := D_Off;
+       --       ^^^^^
+       --  CONSTRAINT_ERROR!
+    end Show_Device;
+
+As we've discussed
+:ref:`previously <Adv_Ada_Record_Discriminants_Object_Assignments>`, when we
+set the values for the discriminants of a type in the object declaration, we're
+constraining the objects. If the discriminants of two objects don't match, the
+:ada:`Constraint_Error` exception is raised at runtime because the
+:ref:`discriminant check <Adv_Ada_Discriminant_Check>` fails. Therefore, in the
+:ada:`Show_Device` procedure, because :ada:`D_Off` and :ada:`D_On` are
+constrained and have different values for the :ada:`State` discriminant, we
+cannot assign them to each other. In contrast, because :ada:`D` wasn't
+constrained at its declaration, we can assign objects with different
+discriminants (such as :ada:`D_Off` and :ada:`D_On`) to it.
+
+Note that the variant part of a record can be more complex. For example, we
+could have an additional discriminant and use it in the variant part:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Device_Boost
+
+    package Devices is
+
+       type Device_State is
+         (Off, On);
+
+       type Device_Info is
+       record
+          V : Float;
+       end record;
+
+       type Device (State : Device_State;
+                    Boost : Boolean) is
+       record
+          case State is
+             when Off =>
+                null;
+             when On =>
+                Info : Device_Info;
+                case Boost is
+                   when False =>
+                      null;
+                   when True =>
+                      Factor : Float;
+                end case;
+          end case;
+       end record;
+
+    end Devices;
+
+In this version of the :ada:`Devices` package, we introduced a *boost button*
+as a discriminant (:ada:`Boost`) and an associated boost factor component
+(:ada:`Factor`) in the variant part.
+
+In the remaining of this section, we discuss a couple of details about variant
```

Comment:
"In the remaining of this section" should say what the word "remaining" references. 
Maybe "In the remaining parts of this section" ?

Reply:
Yes, "parts" is missing here.